File: cpp-std/demo00.cpp
/*
INTRODUCTION TO MULTITHREADING
You should try running this app several times and see results.
*/


#include <iostream>
#include <thread>
using namespace std;



void threadFunc() {
    for (int i = 0; i < 300; ++i)
        cout << "B";
}



int main() {
    std::thread th(threadFunc);

    for (int i = 0; i < 300; ++i)
        cout << "A";

    th.join();

    cout << endl;
    return 0;
}

File: js-nodejs/demo00.js
/*
INTRODUCTION TO MULTITHREADING
You should try running this app several times and see results.
*/

import * as mylib from './mylib.js';
import { Worker, isMainThread } from 'worker_threads';


const workerFunc = async () => {
  for (let i = 0; i < 300; ++i) {
    await mylib.sleep(1);
    process.stdout.write('B');
  }
};


const mainFunc = async () => {
  const worker = new Worker(new URL(import.meta.url));

  for (let i = 0; i < 300; ++i) {
    await mylib.sleep(1);
    process.stdout.write('A');
  }
};


if (isMainThread) {
  await mainFunc();
} else {
  await workerFunc();
}
